File: src/main/java/com/jore/epoc/bo/orders/AdjustCreditLineOrder.java
package com.jore.epoc.bo.orders;

import org.hibernate.annotations.CompositeType;
import org.hibernate.annotations.Type;

import com.jore.datatypes.money.Money;
import com.jore.datatypes.percent.Percent;
import com.jore.epoc.bo.accounting.FinancialAccounting;
import com.jore.epoc.bo.message.MessageLevel;

import jakarta.persistence.AttributeOverride;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;

@Entity
public class AdjustCreditLineOrder extends AbstractSimulationOrder {
    private CreditEventDirection direction;
    @AttributeOverride(name = "amount", column = @Column(name = "adjust_amount"))
    @AttributeOverride(name = "currency", column = @Column(name = "adjust_currency"))
    @CompositeType(com.jore.datatypes.hibernate.MoneyCompositeUserType.class)
    private Money amount;
    @Type(com.jore.datatypes.hibernate.PercentUserType.class)
    private Percent interestRate;

    @Override
    public void execute() {
        if (direction.equals(CreditEventDirection.INCREASE)) {
            book(getExecutionMonth().atDay(FIRST_OF_MONTH), "Increase credit line by " + amount, FinancialAccounting.BANK, FinancialAccounting.LONG_TERM_DEBT, amount);
            addMessage(MessageLevel.INFORMATION, "CreditLineIncreased", amount, getExecutionMonth());
            setExecuted(true);
        } else if (direction.equals(CreditEventDirection.DECREASE)) {
            if (getCompany().getAccounting().checkFunds(amount, getExecutionMonth().atEndOfMonth())) {
                book(getExecutionMonth().atDay(FIRST_OF_MONTH), "Decrease credit line by " + amount, FinancialAccounting.LONG_TERM_DEBT, FinancialAccounting.BANK, amount);
                addMessage(MessageLevel.INFORMATION, "CreditLineDecreased", amount, getExecutionMonth());
                setExecuted(true);
            } else {
                addMessage(MessageLevel.WARNING, "InsufficientFundToDecrase", amount, getCompany().getAccounting().getBankBalance(getExecutionMonth().atEndOfMonth()));
            }
        }
    }

    @Override
    public Money getAmount() {
        return amount;
    }

    @Override
    public int getSortOrder() {
        return 1;
    }

    @Override
    public String getType() {
        return direction + " Credit Line";
    }

    public void setAmount(Money amount) {
        this.amount = amount;
    }

    public void setDirection(CreditEventDirection direction) {
        this.direction = direction;
    }

    public void setInterestRate(Percent interestRate) {
        this.interestRate = interestRate;
    }
}

File: EpocServer/Sources/App/Models/Orders/AdjustCreditLineOrder.swift
import Foundation

public final class AdjustCreditLineOrder: SimulationOrder {
    public var id: Int64?
    public var executionMonth: YearMonth = .of(2020, 1)
    public var isExecuted: Bool = false
    public weak var company: Company!
    public var direction: CreditEventDirection = .increase
    public var amount: Money = Money.of("CHF", 0)
    public var interestRate: Percent = Percent.of("5%")

    public var sortOrder: Int { 1 }
    public var type: String { "\(direction) Credit Line" }

    public func getAmount() -> Money { amount }

    public func execute() {
        if direction == .increase {
            book(bookingDate: executionMonth.atDay(1), bookingText: "Increase credit line by \(amount)",
                 debitAccount: FinancialAccounting.BANK, creditAccount: FinancialAccounting.LONG_TERM_DEBT, amount: amount)
            addMessage(.information, "CreditLineIncreased", amount, executionMonth)
            isExecuted = true
        } else if direction == .decrease {
            if company.accounting.checkFunds(amount, valueDate: executionMonth.atEndOfMonth()) {
                book(bookingDate: executionMonth.atDay(1), bookingText: "Decrease credit line by \(amount)",
                     debitAccount: FinancialAccounting.LONG_TERM_DEBT, creditAccount: FinancialAccounting.BANK, amount: amount)
                addMessage(.information, "CreditLineDecreased", amount, executionMonth)
                isExecuted = true
            } else {
                addMessage(.warning, "InsufficientFundToDecrase", amount, company.accounting.getBankBalance(executionMonth.atEndOfMonth()))
            }
        }
    }
}
